### ISSUE TITLE
Error in generated file: "Cannot assign through subscript: subscript is get-only"

### ISSUE THREAD

**Issue by logicxd** [birdrides/mockingbird#281]
## New Issue Checklist

- [x] I updated the framework and generator to the latest version
- [x] I searched the [existing GitHub issues](https://github.com/birdrides/mockingbird/issues) and list of [common problems](https://mockingbirdswift.com/common-problems)

## Overview

Protocols that has a subscript as GET only will generate the file that fails to build with error "Cannot assign through subscript: subscript is get-only"

## Example

Ran script: 

```
mockingbird generate --targets "PrivateFrameworkDependency" --project ../../PrivateFrameworkDependency/PrivateFrameworkDependency.xcodeproj --output-dir ./MockingbirdMocks
```

Can reproduce by: 

![CleanShot 2022-01-26 at 00 26 24](https://user-images.githubusercontent.com/12219300/151128742-fd806d6a-b2cf-4dc5-bd57-e10976903355.png)

Which generates: 

![CleanShot 2022-01-26 at 00 05 57](https://user-images.githubusercontent.com/12219300/151128777-e5adb6f1-5bf2-4d1c-a7c0-d09eeee64419.png)

```
// MARK: - Mocked SampleProtocol
public final class SampleProtocolMock: VSPersist.SampleProtocol, Mockingbird.Mock {
  typealias MockingbirdSupertype = VSPersist.SampleProtocol
  static let staticMock = Mockingbird.StaticMock()
  public let mockingbirdContext = Mockingbird.Context(["generator_version": "0.19.2", "module_name": "VSPersist"])

  fileprivate init(sourceLocation: Mockingbird.SourceLocation) {
    self.mockingbirdContext.sourceLocation = sourceLocation
    SampleProtocolMock.staticMock.mockingbirdContext.sourceLocation = sourceLocation
  }

  // MARK: Mocked subscript<T>(`key`: String)
  public subscript<T>(`key`: String) -> T? {
    get {
      return self.mockingbirdContext.mocking.didInvoke(Mockingbird.SwiftInvocation(selectorName: "get.subscript<T>(`key`: String) -> T?", setterSelectorName: "set.subscript<T>(`key`: String, `newValue`: T?) -> T?", selectorType: Mockingbird.SelectorType.subscriptGetter, arguments: [Mockingbird.ArgumentMatcher(`key`)], returnType: Swift.ObjectIdentifier((T?).self))) {
        self.mockingbirdContext.recordInvocation($0)
        let mkbImpl = self.mockingbirdContext.stubbing.implementation(for: $0)
        if let mkbImpl = mkbImpl as? (String) -> T? { return mkbImpl(`key`) }
        if let mkbImpl = mkbImpl as? () -> T? { return mkbImpl() }
        if let mkbImpl = mkbImpl as? (Any?) -> Any { return Mockingbird.dynamicCast(mkbImpl(`key` as Any?)) as T? }
        if let mkbImpl = mkbImpl as? () -> Any { return Mockingbird.dynamicCast(mkbImpl()) as T? }
        for mkbTargetBox in self.mockingbirdContext.proxy.targets(for: $0) {
          switch mkbTargetBox.target {
          case .super:
            break
          case .object(let mkbObject):
            guard var mkbObject = mkbObject as? MockingbirdSupertype else { break }
            let mkbValue: T? = mkbObject[`key`]
            self.mockingbirdContext.proxy.updateTarget(&mkbObject, in: mkbTargetBox)
            return mkbValue
          }
        }
        if let mkbValue = self.mockingbirdContext.stubbing.defaultValueProvider.value.provideValue(for: (T?).self) { return mkbValue }
        self.mockingbirdContext.stubbing.failTest(for: $0, at: self.mockingbirdContext.sourceLocation)
      }
    }
    set {
      return self.mockingbirdContext.mocking.didInvoke(Mockingbird.SwiftInvocation(selectorName: "set.subscript<T>(`key`: String, `newValue`: T?) -> T?", setterSelectorName: "set.subscript<T>(`key`: String, `newValue`: T?) -> T?", selectorType: Mockingbird.SelectorType.subscriptSetter, arguments: [Mockingbird.ArgumentMatcher(`key`), Mockingbird.ArgumentMatcher(newValue)], returnType: Swift.ObjectIdentifier(Void.self))) {
        self.mockingbirdContext.recordInvocation($0)
        let mkbImpl = self.mockingbirdContext.stubbing.implementation(for: $0)
        if let mkbImpl = mkbImpl as? (String, T?) -> Void { return mkbImpl(`key`, newValue) }
        if let mkbImpl = mkbImpl as? () -> Void { return mkbImpl() }
        if let mkbImpl = mkbImpl as? (Any?, Any?) -> Any { return Mockingbird.dynamicCast(mkbImpl(`key` as Any?, newValue as Any?)) as Void }
        if let mkbImpl = mkbImpl as? () -> Any { return Mockingbird.dynamicCast(mkbImpl()) as Void }
        for mkbTargetBox in self.mockingbirdContext.proxy.targets(for: $0) {
          switch mkbTargetBox.target {
          case .super:
            break
          case .object(let mkbObject):
            guard var mkbObject = mkbObject as? MockingbirdSupertype else { break }
            let mkbValue: Void = mkbObject[`key`] = newValue
            self.mockingbirdContext.proxy.updateTarget(&mkbObject, in: mkbTargetBox)
            return mkbValue
          }
        }
        if let mkbValue = self.mockingbirdContext.stubbing.defaultValueProvider.value.provideValue(for: (Void).self) { return mkbValue }
        self.mockingbirdContext.stubbing.failTest(for: $0, at: self.mockingbirdContext.sourceLocation)
      }
    }
  }

  public func getSubscript<T>(_ `key`: @autoclosure () -> String) -> Mockingbird.Mockable<Mockingbird.SubscriptGetterDeclaration, (String) -> T?, T?> {
    return Mockingbird.Mockable<Mockingbird.SubscriptGetterDeclaration, (String) -> T?, T?>(mock: self, invocation: Mockingbird.SwiftInvocation(selectorName: "get.subscript<T>(`key`: String) -> T?", setterSelectorName: "set.subscript<T>(`key`: String, `newValue`: T?) -> T?", selectorType: Mockingbird.SelectorType.subscriptGetter, arguments: [Mockingbird.resolve(`key`)], returnType: Swift.ObjectIdentifier((T?).self)))
  }

  public func setSubscript<T>(_ `key`: @autoclosure () -> String, `newValue`: @autoclosure () -> T?) -> Mockingbird.Mockable<Mockingbird.SubscriptSetterDeclaration, (String, T?) -> Void, Void> {
    return Mockingbird.Mockable<Mockingbird.SubscriptSetterDeclaration, (String, T?) -> Void, Void>(mock: self, invocation: Mockingbird.SwiftInvocation(selectorName: "set.subscript<T>(`key`: String, `newValue`: T?) -> T?", setterSelectorName: "set.subscript<T>(`key`: String, `newValue`: T?) -> T?", selectorType: Mockingbird.SelectorType.subscriptSetter, arguments: [Mockingbird.resolve(`key`), Mockingbird.resolve(`newValue`)], returnType: Swift.ObjectIdentifier(Void.self)))
  }
}

/// Returns a concrete mock of `SampleProtocol`.
public func mock(_ type: VSPersist.SampleProtocol.Protocol, file: StaticString = #file, line: UInt = #line) -> SampleProtocolMock {
  return SampleProtocolMock(sourceLocation: Mockingbird.SourceLocation(file, line))
}
```

## Expected Behavior

There should be no build errors. Best case scenario, it should allow mocking the protocol. Worst case scenario, throw a descriptive run-time error so at least we know its not supported. 

## Environment

* Mockingbird CLI version (`mockingbird version`)
  * 0.19.2
* Xcode and Swift version (`swift --version`): 
  * swift-driver version: 1.26.21 Apple Swift version 5.5.2 (swiftlang-1300.0.47.5 clang-1300.0.29.30)
  * Target: x86_64-apple-macosx11.0
* Package manager (CocoaPods, Carthage, SPM project, SPM package)
  * CocoaPods
* Unit testing framework (XCTest, Quick/Nimble)
  * XCTest
* Custom configuration
  - [ ] Mockingbird ignore files
  - [ ] Supporting source files

## Other Notes

This project uses multiple private frameworks as dependencies. As a result, using `configure` didn't seem to work since it can only link 1 framework at a time. To get around this, I generated the scripts manually using `generate` and was planning to add them to the build-phases. Downside of using build-phases is that it generates classes for EVERYTHING even when not used. 

As a temporary workaround, we could try removing file from getting picked up in the `.mockingbird-ignore` but is not ideal.

## Alternatively

Would be nice to be able to run `configure` for multiple frameworks so that in the added build-phase, there are multiple lines of `generate` calls to handle everything for us

**Comment by andrewchang-bird** [comment]
Thanks for reporting this, Aung.

> Would be nice to be able to run configure for multiple frameworks

You can pass `--preserve-existing` to the configurator to add multiple build phases in succession. I recommend this approach over creating a monolithic run script build phase.

```console
$ mockingbird configure MyTestTarget -- --targets MyLibrary --project /path/to/MyLibrary.xcodeproj
$ mockingbird configure MyTestTarget --preserve-existing -- --targets MyOtherLibrary --project /path/to/MyOtherLibrary.xcodeproj
```

Let me know if that works for you. Otherwise in your custom scripts (not recommended if you’re using CocoaPods) you can pass `--test bundle` to enable thunk pruning; see the SPM package guide for an example there.

**Comment by logicxd** [comment]
Such a quick response :D. This worked! I didn't see the `--preserve-existing` flag while reading the documentations.

One more problem came up though. It works fine in command line but if I were to set it up using build phases, the `configure` command will stop the build since it seems to modify the xcode build phases. Things I tried: 
* Create new target purely for running MockingBird (same issue)
* Use pre-action script for the new target (same issue)

So pretty much, I can't automate it to run on xcode build but would have to run in command line. Would you know a way around this? I'll try `--testbundle` next

**Comment by andrewchang-bird** [comment]
The configurator is meant to just be a one-time action that sets up a project’s build phases. Can you elaborate on why you need to run the configurator automatically as part of a build phase? Does the CocoaPods [`post_install`](https://guides.cocoapods.org/syntax/podfile.html#post_install) hook work for your use case?

> Unfortunate, the `--testbundle` approach did not prune properly.

You might need to also set `SRCROOT` for 0.19. It’ll be inferred based on the environment project in the next release. I’d recommend adding `--verbose` when running the generator manually to see why it’s not choosing to prune thunks.

**Comment by logicxd** [comment]
Ah I see. I wrongly thought the configurator was needed to run after every new model changes since it seems to have some optimizations done. 

With that, I'm unblocked and can get build to work. Thanks for your time and responses. Will let you close this issue
